Dali proxy2 (#5726)
NVIDIA/DALI: docs/plugins/pytorch_dali_proxy.rst

PyTorch DALI Proxy
==================

> Overview

Reviewer: Perhaps even in the overview, I'd like to spare the users some disillusionment and state explicitly that DALI cannot be followed by any further processing steps.

**1. DALI Pipeline**

The DALI pipeline defines the data processing steps. Input data is fed using ``fn.external_source``.

.. code-block:: python

   from nvidia.dali import pipeline_def, fn, types

   @pipeline_def
   def example_pipeline():
       images = fn.external_source(name="images", no_copy=True)
       images = fn.io.file.read(images)
       images = fn.decoders.image(images, device="mixed", output_type=types.RGB)
       return fn.resize(images, size=[224, 224])

   pipeline = example_pipeline(batch_size=32, num_threads=2, device_id=0)

**2. DALI Server and Proxy**

The DALI Server manages the execution of the pipeline. The Proxy acts as an interface for PyTorch data workers.

.. code-block:: python

   from nvidia.dali.plugin.pytorch.experimental import proxy as dali_proxy

   with dali_proxy.DALIServer(pipeline) as dali_server:
      # ... in the workers
      transform_fn = dali_server.proxy  # A callable interface for workers
      future_samples = [transform_fn(image) for image in images]

**3. Data Collation and Execution**

The ``default_collate`` function combines processed samples into a batch. DALI executes the pipeline asynchronously when a batch is collated.

> This step is usually abstracted away inside the PyTorch DataLoader and the user doesn't need to take care of it explicitly.

Reviewer: Swap the paragaphs.

PyTorch DALI Proxy
==================

Overview
--------

**DALI Proxy** is a tool designed to integrate NVIDIA DALI pipelines with PyTorch data workers while maintaining the simplicity of PyTorch's dataset logic. The key features of DALI Proxy include:

> - **Efficient GPU Utilization**: DALI Proxy ensures GPU data processing occurs in the same process running the main loop. This avoids performance degradation caused by multiple CUDA contexts for the same GPU.

Reviewer: ```suggestion
- **Efficient GPU Utilization**: DALI Proxy ensures GPU data processing occurs in the process running the main loop. This avoids performance degradation caused by multiple CUDA contexts for the same GPU.
```

Author: Rewrote that

- **Selective Offloading**: Users can offload parts of the data processing pipeline to DALI while retaining PyTorch Dataset logic, making it ideal for multi-modal applications.

This tutorial will explain the key components, workflow, and usage of DALI Proxy in PyTorch.

.. note::

>    **Disclaimer**: Currently, DALI proxy can only be used in data loaders if no additional processing is required on the outputs provided by DALI before the data is consumed in the training loop.

Reviewer: ```suggestion
   **Disclaimer**: Currently, DALI proxy can only be used in data loaders where DALI outputs need no further processing and are consumed directly by the main loop.
```
1. I think it reads better (subjective)
2. Training loop -> main loop (we can use it for validation, etc).

Reviewer 2: I still struggle to read both variants. Can we maybe say that "It is not possible to process the data inside the dataset between returning it from DALI Proxy and consumption by the training loop"?

Author: Rewrote it

       return fn.resize(images, size=[224, 224])

   pipeline = example_pipeline(batch_size=32, num_threads=2, device_id=0)

**2. DALI Server and Proxy**

The :class:`nvidia.dali.plugin.pytorch.experimental.proxy.DALIServer` manages the execution of the pipeline. The Proxy acts 
as an interface for PyTorch data workers.
Note that the DALI pipeline should contain at least one input (an :meth:`~nvidia.dali.fn.external_source` instance), and
that the names of those nodes then become the inputs to the DALI proxy callable.

.. code-block:: python

   from nvidia.dali.plugin.pytorch.experimental import proxy as dali_proxy
   with dali_proxy.DALIServer(pipeline) as dali_server:
      future_samples = [dali_server.proxy(image) for image in images]

With more than one input, we can choose to use positional arguments, keyword arguments:

.. code-block:: python

   import numpy as np
   from nvidia.dali import pipeline_def, fn, types
   from nvidia.dali.plugin.pytorch.experimental import proxy as dali_proxy

   @pipeline_def
   def example_pipeline2(device):
      a = fn.external_source(name="a", no_copy=True)
      b = fn.external_source(name="b", no_copy=True)

>       return a + b, b - a

Reviewer: ```suggestion
      return b - a, a + b
```
Kaboom, now the pipeline input names are `b', `'a'` and the order of positional arguments is swapped.
